### ISSUE TITLE
# FIXME: this is a quick and dirty solution to Filename too long

### ISSUE THREAD

**Issue by fredharvey** [coursera-dl/coursera-dl#665]
### Subject of the issue
File Paths are too long for some courses (Windows). While Windows 10 allows you to extend beyond 260 characters many programs such as video players, Adobe Acrobat and others don't understand and barf.

note: added a tweak to keep file names shorted in my local script, this helps a little. It used to be 200 (define.py)
TITLE_MAX_LENGTH = 40 

This is an actual file path to a file and its pretty absurd and redundant. Since I'm not reading the folders like a book but using them as a navigational tool to get to content I don't need the full name (the sequence helps more than anything). 

".\Spec. - Entrepreneurship & Strategic Innovation\strategic-innovation-building-and-sustaining-innovative-organizations\05_module-4-executing-innovative-business-models-the-whole-is-bigger-than-the-sum\03_lesson-4-2-elements-of-a-business-model-and-business-model-canvas\03_lesson-4-2-elements-of-a-business-model-and-business-model-canvas.m3u"

QUESTIONS:
- where does the folder name get created? 
- can I just change formatting.py to truncate lecname AND title?

i.e. (using constants of course)
lecname[:20]
title[:40]

### Your environment
* Operating System (name/version): Windows 10
* Python version: 365
* coursera-dl version: 0.11.2

### Steps to reproduce
coursera-dl strategic-innovation-innovation-at-the-frontier

* Is the problem happening with the latest version of the script? yes
* Do you have all the recommended versions of the modules? See them in the
  file `requirements.txt`. yes
* What is the course that you are trying to access? strategic-innovation-innovation-at-the-frontier
* What is the precise command line that you are using (don't forget to obfuscate
  your username and password, but leave all other information untouched).
* What are the precise messages that you get? Please, use the `--debug`
  option before posting the messages as a bug report. Please, copy and paste
  them.  Don't reword/paraphrase the messages.

### Expected behaviour
Ideally folder names should indicate sequence (01, 02, 03, etc) and a subset of the name (maybe 10-15 characters) and the full name of the module/section whatever could be a simple text file in the folder.

### Actual behaviour
very long folder and file names

**Comment by balta2ar** [comment]
This is a known problem we haven't had a chance to get our hands on. You're right about the location: formatting.py is a good place to start.

> Ideally folder names should indicate sequence (01, 02, 03, etc) and a subset of the name (maybe 10-15 characters) and the full name of the module/section whatever could be a simple text file in the folder

That's not as straightforward as it looks once you get into the details. First, it looks like Linux users do not need the truncation at all. Should we enable truncation for Windows users only? How do we transition to this change? Do we just turn it on for thousands of our users and ruin their current folder structure making them download same content twice? Should there be a flag? There is plenty of them already and it's a maintenance burden to support another one. Besides, why 10-15 characters? What if course authors named their chapters so that there is common prefix in each chapter which makes it useless to leave only 15 characters? Should we dynamically trim characters? Well, that sounds a bit complicated.

I'm not expecting that you answer this questions, I'm just suggesting that it's not trivial once you start thinking about the details. And it's the details that matter.

**Comment by fredharvey** [comment]
i agree its not a simple problem but like getting the conversation started - thank you :)

I'd also agree using a flag to enable the different behavior is ideal so it doens't break people who don't need it.

I'm happy to just do a local hack to get things working but I can't see where folder names are created. I've tried changing the functions in the formatting.py but folders are created in a very inconsistent manner

**Comment by balta2ar** [comment]
I think `section_dir` is what you're looking for:
```
ipdb> section_dir
'./google-machine-learning/01_introduction-to-specialization/01_introduction'
```

**Comment by fredharvey** [comment]
interesting so it would appear the hierarchy is 
   portfolio/specialization/class/module/section/lecture/resource


 for example
     iMBA/strategic-leadership/everyday-leadership-foundation/modules../sections.../lectures.../resources...

while the code currently doesn't handle portfolios or specializations like understanding the whole picture helps more. 
https://www.coursera.org/degrees/imba
https://www.coursera.org/specializations/strategic-leadership

**Comment by fredharvey** [comment]
i love the specialization download feature. thank you !

**Comment by fredharvey** [comment]
Created a Windows Forms app (C#) for editing the syllabus 

https://github.com/fredharvey/CourseraSyllabusEditor

Simple application will allow you to open a syllabus file, edit all the contents (simply) and write back out the JSON.

**Comment by skrilmps** [comment]
I am having the same issue with filenames that are too long.

I went into define.py and changed TITLE_MAX_LENGTH to 40, then re-downloaded the course, and the filenames didn't appear to change at all.
